icu4j/main/classes/core/src/com/ibm/icu/impl/units/MeasureUnitImpl.java

```diff
@@ -0,0 +1,740 @@
+            // Dimensionless, constructed by the default constructor: no appending
+            // to this.result, we wish it to contain the zero-length string.
+            return;
+        }
+        if (this.complexity == MeasureUnit.Complexity.COMPOUND) {
+            // Note: don't sort a MIXED unit
+            Collections.sort(this.getSingleUnits(), new SingleUnitComparator());
+        }
+
+        StringBuilder result = new StringBuilder();
+        boolean beforePre = true;
+        boolean firstTimeNegativeDimension = false;
+        for (SingleUnitImpl singleUnit :
+                this.getSingleUnits()) {
+            if (beforePre && singleUnit.getDimensionality() < 0) {
+                beforePre = false;
+                firstTimeNegativeDimension = true;
+            } else if (singleUnit.getDimensionality() < 0) {
+                firstTimeNegativeDimension = false;
+            }
+
+            String singleUnitIdentifier = singleUnit.getNeutralIdentifier();
+            if (this.getComplexity() == MeasureUnit.Complexity.MIXED) {
+                if (result.length() != 0) {
+                    result.append("-and-");
+                }
+            } else {
+                if (firstTimeNegativeDimension) {
+                    if (result.length() == 0) {
+                        result.append("per-");
+                    } else {
+                        result.append("-per-");
+                    }
+                } else {
+                    if (result.length() != 0) {
+                        result.append("-");
+                    }
+                }
+            }
+
+            result.append(singleUnitIdentifier);
+        }
+
+        this.identifier = result.toString();
+    }
+
+    class SingleUnitComparator implements Comparator<SingleUnitImpl> {
+        @Override
+        public int compare(SingleUnitImpl o1, SingleUnitImpl o2) {
+            return o1.compareTo(o2);
+        }
+    }
+
+    public static class UnitsParser {
+        public static MeasureUnitImpl parseForIdentifier(String identifier) {
+            UnitsParser parser = new UnitsParser(identifier);
+            return parser.parse();
+
+        }
+
+        private MeasureUnitImpl parse() {
+            MeasureUnitImpl result = new MeasureUnitImpl();
+
+            if (fSource.isEmpty()) {
+                // The dimensionless unit: nothing to parse. return the empty result.
+                return result;
+            }
+
+
+            int unitNum = 0;
+            while (hasNext()) {
+                sawAnd = false;
+                SingleUnitImpl singleUnit = nextSingleUnit();
+                Assert.assrt(!singleUnit.isDimensionless());
+
+                boolean added = result.appendSingleUnit(singleUnit);
+                if (sawAnd && !added) {
+                    throw new InternalError("Two similar units are not allowed in a mixed unit.");
+                }
+
+                if ((result.singleUnits.size()) >= 2) {
```

Comment:
this something I want to discuss, 

is `meter-square-meter` is going to be compound ??

this weird, because in this case `getSingleUnit` will return single unit and the single unit list will equal to 1.

I feel it should be single and not compound

Reply:
The string `"meter-square-meter"` should get normalized to `"cubic-meter"` as soon as it is parsed.

Reply:
okay, this means `meter-square-meter` is a SingleUnit and not a Compound one.

Reply:
Done!
https://github.com/icu-units/icu/issues/87

Reply:
I don't see a meter-square-meter unit test.

To me personally: I get a really good and positive feeling out of seeing a unit test that failed end up passing after I implement the fix. It makes what is otherwise a "theoretical fix" feel a lot more real. (Carrot!)

Reply:
I added to the issue that we need to add test case for it

Reply:
Great, thanks!